Add support for INDEX BY an associated entity

andrews05 opened the pull request:
This allows specifying an association in the INDEX BY clause of a query
which will index by the association's join column.

Related to #7661.

Note I haven't added support for specifying the name of a related entity in an indexed association ("indexBy" annotation) as I couldn't see a good way of doing so. Using the name of the join column does work here though, so it may not be necessary.

beberlei commented:
I like this very much

beberlei commented on lib/Doctrine/ORM/Query/SqlWalker.php:
```diff
@@ -883,9 +883,39 @@ public function walkIdentificationVariableDeclaration($identificationVariableDec
+
+                $joinColumns = $association->getJoinColumns();
+
+                // COMPOSITE KEYS NOT (YET?) SUPPORTED
```
they can't be supported for this, so its ok to remove the comment. The exception is exactly showing users this is not supported.

andrews05 replied:
The switch here is copied from walkPathExpression(), including the comments. Happy to remove the comment if you like though.

beberlei replied:
Yes i guess removing the comment makes sense then, also at the other location.

beberlei commented:
@andrews05 can you make the PR open against the 2.8 branch instead? This is not a feature that requires to wait for 3.0 release

beberlei commented:
You should also add a sentence to the index by section in this doc page that both single value field and assocations are supported: https://github.com/doctrine/orm/blob/master/docs/en/reference/dql-doctrine-query-language.rst

andrews05 commented:
@beberlei I've made the changes suggested (removed comment, note in docs).
I opened another PR #7918 for the 2.8 branch. The code is a little different, so should I keep this one open as well?

andrews05 committed:
Add support for INDEX BY an associated entity

This allows specifying an association in the INDEX BY clause of a query
which will index by the association's join column.

Related to #7661.

beberlei commented:
@lcobucci what do you think w.r.t to merge strategy of this feature? how do you upmerge changes from 2.8 to master?

lcobucci commented:
@beberlei we're going to organise things so that we can merge things up to master more easily. I'd say we can keep this in 2.8.x only.
